apps/bbsync/query.py

```diff
@@ -128,12 +128,29 @@ def generate_component(self):
         """
         generate component to differentiate between flaw and flaw draft
         """
-        self._query["component"] = (
-            "vulnerability-draft"
-            if self.flaw.workflow_state == Flaw.WorkflowState.NEW
-            else "vulnerability"
+
+        # NEW and REJECTED should get "vulnerability-draft", unless
+        # it already has "vulnerability" component.
+        # Otherwise it should get "vulnerability" component.
+        # Full logic explanation in OSIDB-3023.
+
+        new_flaw = self.flaw.workflow_state == Flaw.WorkflowState.NEW
+
+        rejected_flaw = self.flaw.workflow_state == Flaw.WorkflowState.REJECTED
+
+        missing_component = not self.flaw.meta_attr.get("bz_component")
+
+        draft_component = (
+            self.flaw.meta_attr.get("bz_component") == "vulnerability-draft"
         )
 
+        component = "vulnerability"
+
+        if (new_flaw or rejected_flaw) and (missing_component or draft_component):
+            component = "vulnerability-draft"
```

Comment:
Not saying you have to change this, but we can reduce some lines of code
```suggestion
        missing_component = not self.flaw.meta_attr.get("bz_component")
        draft_component = (
            self.flaw.meta_attr.get("bz_component") == "vulnerability-draft"
        )
        component = "vulnerability-draft" if (
            (self.flaw.workflow_state in {Flaw.WorkflowState.NEW, Flaw.WorkflowState.REJECTED}) and
            (missing_component or draft_component))
        ) else "vulnerability"
```

Reply:
Thanks. Is it acceptable now?